torch/__init__.py

```diff
@@ -1158,11 +1158,23 @@ def _check_is_size(i, message=None):
     to infer the size from context, or if you should wrap-around or truncate).
     Only use this if the only valid value is an honest to goodness size.
     """
+    from torch.fx.experimental.sym_node import SymNode
+    import sympy
+    is_sym = isinstance(i, SymInt) and isinstance(i.node, SymNode) and isinstance(i.node.expr, sympy.Symbol)
+    if is_sym:
+        if i.node.expr in i.node.shape_env.check_is_size_called:
+            return
```

Comment:
I think this can be written in a more tight fashion. My primary thought is whether or not the constituent pieces of this (_check and _advise_is_size) can be memoized. In particular, the _check memo would a short circuit for the case when you are testing `i >= 0` to consult value ranges directly before going the slow path. The _advise_is_size could memoize directly on sympy.Symbol, but it's not clear to me that it is that complicated.